### PR TITLE
70 load balancer

### PR DESCRIPTION
**Pull request by LimKianAn**
- `Service LoadBalancer` is created/deleted alongside `Postgres`.
- `Endpoints` are correctly picked up through the labels.
- `nextFreePort()` in *lbmanager.go* works.

@majst01 @eberlep Please comment if the direction is correct.

### CODE REVIEW

**Comment by majst01** [comment]
LGTM

File: api/v1/postgres_types.go [hunk 1]
```diff
@@ -136,6 +138,14 @@ type PostgresStatus struct {
 	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 	Description string `json:"description,omitempty"`
+
+	LBSocket LBSocket `json:"lbSocket,omitempty"`
```
Comment (majst01) [inline api/v1/postgres_types.go:142]: I would prefer Socket instead of LBSocket

File: charts/postgreslet/values.yaml [hunk 2]
```diff
@@ -66,6 +66,12 @@ postgreslet:
   tenant: sample-tenant
   # metricsAddr defines the listen address of the metrics endpoint
   metricsAddr: ":8080"
+  # loadBalancerIP defines the load-balancer IP of postgres in this cluster
+  loadBalancerIP: "172.0.0.1"
+  # portRangeStart deinfes the start of the port range of services LoadBalancer
+  portRangeStart: 31767
```
Comment (eberlep) [inline charts/postgreslet/values.yaml:72]: Which ports to use as defaults?

File: api/v1/postgres_types.go [hunk 3]
```diff
@@ -205,6 +215,54 @@ func (p *Postgres) ToKey() *types.NamespacedName {
...
+
+	lb.Namespace = p.ToPeripheralResourceNamespace()
+	lb.Name = p.ToSvcLBName()
+	label := map[string]string{
+		"application":  "spilo",
+		"cluster-name": p.ToPeripheralResourceName(),
+		"spilo-role":   "master",
+		"team":         p.generateTeamID(),
+	}
+	lb.SetLabels(label)
```
Comment (eberlep) [inline api/v1/postgres_types.go:234]: Add our own custom label here to ease selection in `lbmanager`

File: api/v1/postgres_types.go [hunk 4]
```diff
@@ -205,6 +215,54 @@ func (p *Postgres) ToKey() *types.NamespacedName {
...
+	// svc.Spec.LoadBalancerSourceRanges // todo: Do we need to set this?
+
+	port := corev1.ServicePort{}
+	port.Name = "postgresql"
+	port.Port = lbPort
+	port.Protocol = corev1.ProtocolTCP
+	port.TargetPort = intstr.FromInt(5432)
+	lb.Spec.Ports = []corev1.ServicePort{port}
+
+	lb.Spec.Selector = label
```
Comment (eberlep) [inline api/v1/postgres_types.go:245]: Make sure *not* to use our custom label here, as this is the selector for operator-managed pods

File: pkg/lbmanager/lbmanager.go [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+	lb := &corev1.Service{}
+	lb.Namespace = in.ToPeripheralResourceNamespace()
+	lb.Name = in.ToSvcLBName()
+	if err := m.Delete(ctx, lb); client.IgnoreNotFound(err) != nil { // todo: remove ignorenotfound
+		return err
+	}
+	return nil
+}
+
+const SvcTypeIndex = "spec.type"
```
Comment (eberlep) [inline pkg/lbmanager/lbmanager.go:61]: Remove this one and use custom label,

File: pkg/lbmanager/lbmanager.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+func (m *LBManager) nextFreePort(ctx context.Context) (int32, error) {
+	lbs := &corev1.ServiceList{}
+	if err := m.List(ctx, lbs,
+		client.MatchingFields{SvcTypeIndex: string(corev1.ServiceTypeLoadBalancer)},
+		client.MatchingLabels{
```
Comment (eberlep) [inline pkg/lbmanager/lbmanager.go:67]: Only use custom label here. We won't need the MatchinField then either.

File: pkg/operatormanager/operatormanager.go [hunk 7]
```diff
@@ -115,6 +115,20 @@ func (m *OperatorManager) IsOperatorDeletable(ctx context.Context, namespace str
 		m.Log.Info("operator is deletable")
 		return true, nil
 	}
+
```
Comment (eberlep) [inline pkg/operatormanager/operatormanager.go:118]: Add TODO to #83

**Review by eberlep — CHANGES_REQUESTED** [review]

File: charts/postgreslet/values.yaml (hunk 2 above)
Comment (eberlep) [reply charts/postgreslet/values.yaml:72]: @majst01 what do you think?

File: api/v1/postgres_types.go [hunk 8]
```diff
@@ -205,6 +215,57 @@ func (p *Postgres) ToKey() *types.NamespacedName {
 	}
 }
 
+var SvcLoadBalancerLabel = map[string]string{
+	"cloud-service": "pgaas",
```
Comment (eberlep) [inline api/v1/postgres_types.go:219]: hmm, how about `{ "postgres.database.fits.cloud/managed-by": "postgreslet"}`?

**Review by eberlep — APPROVED** [review]

**Pull request merged**